File: src/models/architectures/lstm_modules.py
"""
LSTM building blocks: ResidualLSTM, Attention, TCN
"""
import torch
import torch.nn as nn


class ResidualLSTM(nn.Module):
    """LSTM with residual connections"""
    def __init__(self, input_size, hidden_size, num_layers=2, dropout=0.3, bidirectional=True):
        super().__init__()
        self.lstm = nn.LSTM(
            input_size,
            hidden_size,
            num_layers,
            batch_first=True,
            bidirectional=bidirectional,
            dropout=dropout if num_layers > 1 else 0
        )
        self.bidirectional = bidirectional
        output_size = hidden_size * 2 if bidirectional else hidden_size
        self.residual_fc = nn.Linear(input_size, output_size) if input_size != output_size else nn.Identity()

    def forward(self, x):
        lstm_out, _ = self.lstm(x)
        residual = self.residual_fc(x)
        # Ensure shapes match
        if residual.dim() == 2:
            residual = residual.unsqueeze(1)
        if residual.shape[1] != lstm_out.shape[1]:
            # Repeat or interpolate residual to match sequence length
            residual = residual.repeat(1, lstm_out.shape[1], 1)
        return lstm_out + residual


class MultiHeadAttention(nn.Module):
    """Multi-head attention for time series"""
    def __init__(self, hidden_size, num_heads=4):
        super().__init__()
        self.num_heads = num_heads
        self.head_dim = hidden_size // num_heads

        assert self.head_dim * num_heads == hidden_size, "hidden_size must be divisible by num_heads"

        self.query = nn.Linear(hidden_size, hidden_size)
        self.key = nn.Linear(hidden_size, hidden_size)
        self.value = nn.Linear(hidden_size, hidden_size)
        self.fc_out = nn.Linear(hidden_size, hidden_size)

    def forward(self, x):
        batch_size, seq_len, _ = x.shape

        Q = self.query(x).view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        K = self.key(x).view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
        V = self.value(x).view(batch_size, seq_len, self.num_heads, self.head_dim).transpose(1, 2)

        attention = torch.matmul(Q, K.transpose(-2, -1)) / torch.sqrt(torch.tensor(self.head_dim, dtype=torch.float32))
        attention = torch.softmax(attention, dim=-1)

        out = torch.matmul(attention, V)
        out = out.transpose(1, 2).contiguous().view(batch_size, seq_len, -1)
        return self.fc_out(out)


class TCNBlock(nn.Module):
    """Temporal Convolutional Network block"""
    def __init__(self, input_size, output_size, kernel_size=3, dilation=1, dropout=0.2):
        super().__init__()
        # Use 'same' padding to maintain sequence length
        padding = ((kernel_size - 1) * dilation) // 2
        self.conv1 = nn.Conv1d(input_size, output_size, kernel_size,
                              padding=padding, dilation=dilation)
        self.conv2 = nn.Conv1d(output_size, output_size, kernel_size,
                              padding=padding, dilation=dilation)
        self.relu = nn.ReLU()
        self.dropout = nn.Dropout(dropout)

        # Skip connection
        self.downsample = nn.Conv1d(input_size, output_size, 1) if input_size != output_size else nn.Identity()

    def forward(self, x):
        residual = self.downsample(x)

        # Apply first convolution
        out = self.relu(self.conv1(x))
        out = self.dropout(out)

        # Apply second convolution
        out = self.conv2(out)

        # Ensure shapes match before adding residual
        if out.shape != residual.shape:
            # Crop or pad to match shapes
            if out.shape[2] > residual.shape[2]:
                out = out[:, :, :residual.shape[2]]
            elif out.shape[2] < residual.shape[2]:
                residual = residual[:, :, :out.shape[2]]

        out = self.relu(out + residual)
        return out

File: cpp/src/models/lstm_modules.cpp
// === FILE: cpp/src/models/lstm_modules.cpp ===
#include "trading/models/lstm_modules.hpp"

#include <cmath>

namespace trading::models {

ResidualLSTMImpl::ResidualLSTMImpl(int input_size, int hidden_size, int num_layers,
                                   double dropout, bool bidirectional) {
    auto options = torch::nn::LSTMOptions(input_size, hidden_size)
                       .num_layers(num_layers)
                       .batch_first(true)
                       .bidirectional(bidirectional)
                       .dropout(num_layers > 1 ? dropout : 0.0);
    lstm_ = register_module("lstm", torch::nn::LSTM(options));

    output_size_ = bidirectional ? hidden_size * 2 : hidden_size;
    identity_ = (input_size == output_size_);
    if (!identity_) {
        residual_fc_ = register_module(
            "residual_fc", torch::nn::Linear(input_size, output_size_));
    }
}

torch::Tensor ResidualLSTMImpl::forward(const torch::Tensor& x) {
    // lstm returns (output, (h_n, c_n)); we only need the output sequence.
    auto lstm_out = std::get<0>(lstm_->forward(x));
    torch::Tensor residual = identity_ ? x : residual_fc_->forward(x);
    return lstm_out + residual;
}

MultiHeadAttentionImpl::MultiHeadAttentionImpl(int hidden_size, int num_heads)
    : num_heads_(num_heads), head_dim_(hidden_size / num_heads) {
    TORCH_CHECK(head_dim_ * num_heads == hidden_size,
                "hidden_size must be divisible by num_heads");
    query_ = register_module("query", torch::nn::Linear(hidden_size, hidden_size));
    key_ = register_module("key", torch::nn::Linear(hidden_size, hidden_size));
    value_ = register_module("value", torch::nn::Linear(hidden_size, hidden_size));
    fc_out_ = register_module("fc_out", torch::nn::Linear(hidden_size, hidden_size));
}

torch::Tensor MultiHeadAttentionImpl::forward(const torch::Tensor& x) {
    const auto batch_size = x.size(0);
    const auto seq_len = x.size(1);

    auto reshape_heads = [&](const torch::Tensor& t) {
        // [b, s, hidden] -> [b, heads, s, head_dim]
        return t.view({batch_size, seq_len, num_heads_, head_dim_}).transpose(1, 2);
    };

    auto Q = reshape_heads(query_->forward(x));
    auto K = reshape_heads(key_->forward(x));
    auto V = reshape_heads(value_->forward(x));

    const double scale = std::sqrt(static_cast<double>(head_dim_));
    auto attention = torch::matmul(Q, K.transpose(-2, -1)) / scale;
    attention = torch::softmax(attention, -1);

    auto out = torch::matmul(attention, V);             // [b, heads, s, head_dim]
    out = out.transpose(1, 2).contiguous().view(        // [b, s, hidden]
        {batch_size, seq_len, num_heads_ * head_dim_});
    return fc_out_->forward(out);
}

}  // namespace trading::models
